Blazor Byte Array Interop Support

TanayParikh opened the pull request:
Implements changes discussed in https://github.com/dotnet/aspnetcore/pull/32259.

Fixes: https://github.com/dotnet/aspnetcore/issues/21877

TanayParikh committed:
Blazorserver Byte Array Interop Support

pranavkm commented on src/Components/Server/src/ComponentHub.cs:
```diff
@@ -209,6 +209,17 @@ public async ValueTask EndInvokeJSFromDotNet(long asyncHandle, bool succeeded, s
+                return;
+            }
+
+            _ = circuitHost.SupplyByteArray(id, data);
```
Any reason not to await this?

TanayParikh replied:
We don't currently await the other interop methods (`BeginInvokeDotNetFromJS`, `EndInvokeJSFromDotNet`) so just following the same fire & forget convention here.

TanayParikh replied:
I'm open to awaiting if you'd think it'd be beneficial?

pranavkm replied:
I guess if we're on an async code path, it makes sense to await it. It's not like the client expects this to be fire-and-forget, is it?

pranavkm commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,92 @@
+{
+    internal sealed class ByteArrayJsonConverter : JsonConverter<byte[]>
+    {
+        /// Atomically incrementing unique identifier for serializing byte arrays.
```
```suggestion
        // Atomically incrementing unique identifier for serializing byte arrays.
```

pranavkm commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -292,6 +294,32 @@ public static void EndInvokeJS(JSRuntime jsRuntime, string arguments)
+            {
+                throw new ArgumentOutOfRangeException($"Element id '${id}' cannot be added to the byte arrays to be revived with length '${jsRuntime.ByteArraysToBeRevived.Count}'.");
+            }
+            else if (data.Length + jsRuntime.ByteArraysToBeRevivedByteLength > (31*1024)) // TODO; get this limit from SignalR somehow?
```
```suggestion
            else if (data.Length - 32k >  jsRuntime.ByteArraysToBeRevivedByteLength) // TODO; get this limit from SignalR somehow?
```

To avoid int overflows.

TanayParikh replied:
I believe this should be:

```suggestion
            else if (32k - data.Length <  jsRuntime.ByteArraysToBeRevivedByteLength) // TODO; get this limit from SignalR somehow?
```

pranavkm commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -222,6 +222,8 @@ private static void EndInvokeDotNetAfterTask(Task task, JSRuntime jsRuntime, in
                 throw new JsonException($"Unexpected JSON token {reader.TokenType}. Ensure that the call to `{methodIdentifier}' is supplied with exactly '{parameterTypes.Length}' parameters.");
             }
 
+            jsRuntime.ResetByteArraysToBeRevived();
```
Does the byte array leak if we encounter any of the errors thrown before this?

TanayParikh replied:
No harm in moving this up before these two checks (to line 213). Even if this remains uncleared, it'll get cleared on the next message that supplies byte arrays.

pranavkm commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,92 @@
+            // We must copy over the byte array to ensure the data is preserved after we
+            // clear the JSRuntime.ByteArraysToBeRevived buffer.
+            var result = new byte[byteArray.Length];
+            Array.Copy(byteArray, result, byteArray.Length);
```
```suggestion
            byteArray.AsSpan().CopyTo(result);
```

pranavkm commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -292,6 +294,32 @@ public static void EndInvokeJS(JSRuntime jsRuntime, string arguments)
+            }
+            else if (id != (jsRuntime.ByteArraysToBeRevived.Count + 1))
+            {
+                throw new ArgumentOutOfRangeException($"Element id '${id}' cannot be added to the byte arrays to be revived with length '${jsRuntime.ByteArraysToBeRevived.Count}'.");
```
```suggestion
                throw new ArgumentOutOfRangeException($"Element id '{id}' cannot be added to the byte arrays to be revived with length '{jsRuntime.ByteArraysToBeRevived.Count}'.");
```

TanayParikh replied:
Thanks, nice catch!

TanayParikh commented on src/Components/Server/src/BlazorPack/BlazorPackHubProtocolWorker.cs:
```diff
@@ -34,6 +35,20 @@ protected override object DeserializeObject(ref MessagePackReader reader, Type t
+                        return Array.Empty<byte>();
+                    }
+
+                    return bytes.Value.ToArray();
```
Still doing `ToArray` in place of passing through the `ReadOnlySequence<byte>`. 

Concerns with using the `ReadOnlySequence` relate to the fact that it doesn't implicitly cast to `byte[]` so when we're examining method parameters we have a type mistmatch.

Additionally, as Javier mentioned in the other PR we need to verify the lifecycle/ownership of the `ReadOnlySequence`.

SteveSandersonMS replied:
> Concerns with using the ReadOnlySequence relate to the fact that it doesn't implicitly cast to byte[] so when we're examining method parameters we have a type mistmatch.

I think you'd need to change the receiving code (`ComponentHub.SupplyByteArray`) to accept a `ReadOnlySequence<byte>` instead of a `byte[]`, so there wouldn't be a type mismatch.

Agreed on verifying the ownership of the buffer before doing that though.

javiercn replied:
Yes, lets keep it as is for the time being and then we can do a quick check and change afterwards.

TanayParikh commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -292,6 +294,32 @@ public static void EndInvokeJS(JSRuntime jsRuntime, string arguments)
+            }
+
+            jsRuntime.ByteArraysToBeRevived.Append(data);
+            jsRuntime.ByteArraysToBeRevivedByteLength += data.Length;
```
As Javier suggested in the prior PR we may want to restrict the max number of byte arrays by doing something like this:
```suggestion
            jsRuntime.ByteArraysToBeRevivedByteLength += Math.Min(4, data.Length);
```

pranavkm commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,92 @@
+
+        public JSRuntime JSRuntime { get; }
+
+        public override bool CanConvert(Type typeToConvert) =>
```
It's super unusual to see ReadOnlySequences appear in interop APIs. Do we really need to support it?

TanayParikh replied:
Nope not at all, that was a leftover from something I was experimenting. Thanks, removed.

TanayParikh committed:
PR Feedback

TanayParikh committed:
Merge branch 'main' into taparik/byteArrayDirectInterop

TanayParikh commented on src/Components/Web.JS/src/Boot.WebView.ts:
```diff
@@ -4,7 +4,7 @@ import { shouldAutoStart } from './BootCommon';
 import { internalFunctions as navigationManagerFunctions } from './Services/NavigationManager';
 import { setEventDispatcher } from './Rendering/Events/EventDispatcher';
 import { startIpcReceiver } from './Platform/WebView/WebViewIpcReceiver';
-import { sendBrowserEvent, sendAttachPage, sendBeginInvokeDotNetFromJS, sendEndInvokeJSFromDotNet, sendLocationChanged } from './Platform/WebView/WebViewIpcSender';
```
Part of the WASM changes (just to get the project to a buildable state)

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts:
```diff
@@ -267,6 +268,13 @@ export module DotNet {
     endInvokeJSFromDotNet(callId: number, succeeded: boolean, resultOrError: any): void;
+
+    /**
+     * Receives notification that a byte array is available for transfer.
```
Minor detail, but helps with keeping track of which direction these calls are going in, since there are symmetric APIs going the other way too :)

```suggestion
     * Invoked by the runtime to transfer a byte array to .NET.
```

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,91 @@
+
+        public override void Write(Utf8JsonWriter writer, byte[] value, JsonSerializerOptions options)
+        {
+            var id = Interlocked.Increment(ref _byteArrayId);
```
We don't need to think about concurrency here. All operations here run in a single-threaded context. (Blazor uses a sync context to ensure only one thread at a time is running code)

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts:
```diff
@@ -346,6 +354,15 @@ export module DotNet {
+    },
+
+    /**
+     * Receives notification that a byte array is available for transfer.
```
```suggestion
     * Receives notification that a byte array is being transferred from .NET to JS.
```

(same reason as above)

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -292,6 +295,32 @@ public static void EndInvokeJS(JSRuntime jsRuntime, string arguments)
+            {
+                throw new ArgumentOutOfRangeException($"Element id '{id}' cannot be added to the byte arrays to be revived with length '{jsRuntime.ByteArraysToBeRevived.Count}'.");
+            }
+            else if ((31 * 1024) - data.Length < jsRuntime.ByteArraysToBeRevivedByteLength) // TODO; get this limit from SignalR somehow?
```
Inject `IOptions<HubOptions>` and set it to the limit defined [here](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/server/Core/src/HubOptions.cs#L44) when creating the instance in the CircuitFactory

SteveSandersonMS replied:
If you do want to get the message size limit from SignalR, you can add a DI dependency to `RemoteJSRuntime`'s constructor so it receives a `Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.SignalR.HubOptions> HubOptions`, and then read the `.Value.MaximumReceiveMessageSize` from that.

I know it's a bit awkward to get the value into `DotNetDispatcher` because of the layering (it doesn't know about `RemoteJSRuntime`). You could declare some `internal protected virtual` property on `JSRuntime` that defines a `MaximumByteDataPerCall` with default value `int.MaxValue`, overridden in `RemoteJSRuntime`. Or I'm sure you can come up with other ways of doing it.

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts:
```diff
@@ -394,24 +411,28 @@ export module DotNet {
-        return jsObject.getWrappedObject();
-      } else {
-        throw new Error(`JS object instance with ID ${id} does not exist (has it been disposed?).`);
+        return byteArray;
```
Great - thanks for consolidating these revivers together!

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts:
```diff
@@ -430,7 +451,22 @@ export module DotNet {
+      }
+
+      return value;
+    }
```
TBH I find this refactoring slightly unexpected. Would it also work to track `byteArrayIndex` outside the function as a file-level variable (just like `nextJsObjectId`), and avoid the extra layer of function wrapping around `argReplacer`?

SteveSandersonMS replied:
Ah no sorry, I realise why you did this now (to reset the ID to `0` at the start of each serialization). This makes sense!

SteveSandersonMS replied:
I guess it would still be a tiny bit more conventional not to create a new `argReplacer` closure instance on each call, by doing something like:

```ts
let nextbyteArrayIndex: number;

function stringifyArgsOrReturnValue(args: any[] | null) {
    nextbyteArrayIndex = 0;
    return JSON.stringify(args, argReplacer);
}

function argReplacer(...) { ... unchanged ... }
```

[This doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#performance_considerations) seems to back this up, but not really a big deal AFAIK!

TanayParikh replied:
> not to create a new argReplacer closure instance on each call

Ahh great catch, thanks!

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -292,6 +295,32 @@ public static void EndInvokeJS(JSRuntime jsRuntime, string arguments)
+        /// <param name="data">Byte array to be transfered from JS.</param>
+        public static void SupplyByteArray(JSRuntime jsRuntime, long id, byte[] data)
+        {
+            if (id == 0)
+            {
+                // Starting a new transfer, clear out previously stored byte arrays
+                jsRuntime.ResetByteArraysToBeRevived();
+            }
```
I think we can probably clear the buffers earlier, like after the `BeginInvokeDotNet` call has deserialized the arguments. This is important because the list will be released right after we are done using them and we will avoid holding on to them for longer than necessary.

SteveSandersonMS replied:
I think it does also do that already. The call here is only to guarantee we clean up if something went wrong during the last call.

javiercn replied:
Wouldn't doing it on a `finally` block the CircuitHost achieve the same result? If something went wrong with the previous call, we should fail and tear-down the circuit, shouldn't we?

TanayParikh replied:
If we'd like to do it from the circuit host, it'd require us to add `ResetByteArraysToBeRevived` to the public API. Would that be acceptable?

SteveSandersonMS replied:
My personal preference would be not making it part of the public API. Given the discussion at https://github.com/dotnet/aspnetcore/pull/33015#discussion_r639619461 it seems like we're OK keeping the cleanup as-is.

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -292,6 +295,32 @@ public static void EndInvokeJS(JSRuntime jsRuntime, string arguments)
+                throw new ArgumentOutOfRangeException("Exceeded the maximum byte array transfer limit for a call.");
+            }
+
+            jsRuntime.ByteArraysToBeRevived.Append(data);
+            jsRuntime.ByteArraysToBeRevivedByteLength += data.Length;
```
These two statements should be a single operation. (Append should also update the length)

SteveSandersonMS replied:
I don't get this suggestion. `ByteArraysToBeRevived` and `ByteArraysToBeRevivedByteLength` are separate things.

TanayParikh replied:
Append will update the length of `ByteArraysToBeRevived` where length is the number of byte arrays. `ByteArraysToBeRevivedByteLength` tracks the total number of bytes contained in `ByteArraysToBeRevived`. This is done to track whether or not we're within the limits of a single SignalR message.

I've renamed `ByteArraysToBeRevivedByteLength` to `ByteArraysToBeRevivedTotalBytes` to hopefully make this distinction clearer.

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -292,6 +295,32 @@ public static void EndInvokeJS(JSRuntime jsRuntime, string arguments)
+            {
+                throw new ArgumentOutOfRangeException($"Element id '{id}' cannot be added to the byte arrays to be revived with length '{jsRuntime.ByteArraysToBeRevived.Count}'.");
+            }
+            else if ((31 * 1024) - data.Length < jsRuntime.ByteArraysToBeRevivedByteLength) // TODO; get this limit from SignalR somehow?
```
We should not do this check here, it should be done on the JSRuntime

SteveSandersonMS replied:
Arguably it would be nice to move most of the logic from this method inside `JSRuntime`, since it exclusively acts upon state inside the `JSRuntime`.

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,91 @@
+    internal sealed class ByteArrayJsonConverter : JsonConverter<byte[]>
+    {
+        // Atomically incrementing unique identifier for serializing byte arrays.
+        private static long _byteArrayId;
```
Can we make this non-static, and stored per-`JSRuntime` instance?

The issue with it being static is that it discloses information about other users' activities. We try to keep all IDs scoped to be per-connection (or GUIDs) so that people can't infer things about the traffic level on the site, how long the server has been up, etc.

javiercn replied:
Good catch

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -292,6 +295,32 @@ public static void EndInvokeJS(JSRuntime jsRuntime, string arguments)
+        /// <param name="jsRuntime">The <see cref="JSRuntime"/>.</param>
+        /// <param name="id">Identifier for the byte array being transfered.</param>
+        /// <param name="data">Byte array to be transfered from JS.</param>
+        public static void SupplyByteArray(JSRuntime jsRuntime, long id, byte[] data)
```
I think this method must go in the JSRuntime implementation directly instead of here. It's suffering from `feature envy` as described [here](https://moderatemisbehaviour.github.io/clean-code-smells-and-heuristics/general/g14-feature-envy.html)

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,91 @@
+            }
+
+            var byteArray = JSRuntime.ByteArraysToBeRevived.Buffer[byteArrayIndex.Value];
+            if (byteArray is null || byteArray.Length == 0)
+            {
+                return Array.Empty<byte>();
+            }
```
Could you clarify what this block is for? Why would one of the array entries be `null`, and why do we care about normalizing already-allocated empty array instances to be `Array.Empty<byte>()`?

TanayParikh replied:
In the case where [message pack fails to read the byte arrays](https://github.com/dotnet/aspnetcore/pull/33015/files#diff-8bdd71729dca8e301353f4156cdf02e587c99ba34e98d23d8361641428e53130R43) (potentially because the user provided a null byte array).

> why do we care about normalizing already-allocated empty array instances to be Array.Empty<byte>()?

We don't/shouldn't. Thanks for pointing that out. Removed.

TanayParikh replied:
Given the thing you pointed out below, removing this block entirely.

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,91 @@
+                throw new JsonException($"Byte array {byteArrayIndex} not found.");
+            }
+
+            var byteArray = JSRuntime.ByteArraysToBeRevived.Buffer[byteArrayIndex.Value];
+            if (byteArray is null || byteArray.Length == 0)
+            {
+                return Array.Empty<byte>();
+            }
```
Under which circumstances would byteArray be null here?

SteveSandersonMS replied:
Oh this is pretty weird, your comment here is the same as mine posted simultaneously, *and* the one below is parallel in exactly the same way.

TanayParikh replied:
Moving discussion to above comment.

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,91 @@
+            // We must copy over the byte array to ensure the data is preserved after we
+            // clear the JSRuntime.ByteArraysToBeRevived buffer.
+            var result = new byte[byteArray.Length];
+            byteArray.AsSpan().CopyTo(result);
```
I don't get this. What's wrong with using the existing array instances without copying? Clearing the `JSRuntime.ByteArraysToBeRevived` buffer will only clear the array of *references* to arrays, not the arrays inside it (at least as far as I'd guess - haven't yet finished reading the whole implementation).

TanayParikh replied:
You're right, this isn't needed, removed.

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,91 @@
+            // clear the JSRuntime.ByteArraysToBeRevived buffer.
+            var result = new byte[byteArray.Length];
+            byteArray.AsSpan().CopyTo(result);
+            return result;
```
Is this copy really necessary? My understanding is that we would have something like `ArrayBuilder<byte[]>();` to hold on to the byte arrays and that we would only call `Clear` on the ArrayBuilder. Aren't the `byte []` arrays created during the deserialization?

TanayParikh replied:
You're right, this isn't needed, removed, thanks!

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -203,13 +203,16 @@ private static void EndInvokeDotNetAfterTask(Task task, JSRuntime jsRuntime, in
                 var parameterType = parameterTypes[index];
                 if (reader.TokenType == JsonTokenType.StartObject && IsIncorrectDotNetObjectRefUse(parameterType, reader))
                 {
+                    jsRuntime.ResetByteArraysToBeRevived();
```
What do you think about if we *didn't* call this before each of the possible failure cases? I know that means some data will hang around for longer than we need, but (1) it's limited to "max signalr message size", so isn't very much, and (2) always gets cleared out before the next incoming call anyway.

Limiting the cleanup to only happen in the "received ID=0" or "finished receiving" cases means there's less for us to remember to do as we evolve this logic in the future, and clarifies that the only thing we actually rely on is that it gets cleared in those cases.

javiercn replied:
Alternatively, why don't we move this to the callsite (`Hub` and equivalent Wasm callsite) and clear it after every JS -> DotNet call completes? (Whether successfully or with a failure)

It's not clear to me what we win by using `0` as a way to reset the buffer. Afaik, we should never be in a situation where the buffer is not empty and we receive a stream with a `0` sequence id. That would mean that the previous JS interop call didn't complete successfully.

SteveSandersonMS replied:
> That would mean that the previous JS interop call didn't complete successfully.

Yes, this was my suggestion for how we can be certain to auto-clean-up in that situation. I agree that a valid alternative is `try`/`finally` around anything that could have failed the previous time to guarantee it was cleaned up already.

TBH I don't mind either way but it's nice to know that even if we make a mistake with forgetting to clean up somewhere, e.g., after some future refactoring when this point gets forgotten about, things will remain safe and errors will only refer to the original cause rather than knock-on effects like breaking *future* JS interop calls.

javiercn replied:
The only case I can think this makes sense is if/when the connection is lost in the middle of a JS interop call (since now it spans multiple messages). I think based on that, it's fine to have it here, I was just a bit worried that we might introduce a hard to find bug in this area where we unknowingly reset the list by mistake. It might be better to be "explicit" about this by forcing the client to send a "reset" frame (with a -1 id) after a failed JS interop call.

In either case, I don't mind this too much, so I think what we have here is good enough.

SteveSandersonMS replied:
> I think what we have here is good enough.

+1

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,91 @@
+
+            long? byteArrayIndex = null;
+
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
+            {
+                if (reader.TokenType == JsonTokenType.PropertyName)
+                {
+                    if (byteArrayIndex is null && reader.ValueTextEquals(ByteArrayRefKey.EncodedUtf8Bytes))
+                    {
+                        reader.Read();
+                        byteArrayIndex = reader.GetInt64();
+                    }
+                    else
+                    {
+                        throw new JsonException($"Unexpected JSON property {reader.GetString()}.");
+                    }
+                }
+                else
+                {
+                    throw new JsonException($"Unexpected JSON Token {reader.TokenType}.");
+                }
+            }
```
There's a simpler way to do this, which involves defining a private type
```csharp
public struct ByteArrayRef
{
  [JsonPropertyName("__byte[]")]
  public int Id { get; set; }
}
```

And calling `JsonSerializer.Deserialize<ArrayRef>(reader, options)` which saves you from dealing with the low level stuff

SteveSandersonMS replied:
Interesting, cool idea

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:
```diff
@@ -20,10 +20,12 @@ public abstract partial class JSRuntime : IJSRuntime
+        private readonly ConcurrentDictionary<long, IDotNetObjectReference> _trackedRefsById = new();
+        private readonly ConcurrentDictionary<long, CancellationTokenRegistration> _cancellationRegistrations = new();
+
+        internal readonly ArrayBuilder<byte[]> ByteArraysToBeRevived = new();
```
`ArrayBuilder<T>` needs to be disposed otherwise it's nonempty, it won't return its final array to the underlying pool.

Perhaps you'll need to make `JSRuntime` itself also disposable to handle this. I can't immediately anticipate any further complexity that would cause but if it did end up being problematic we could think about ways to not use a pooled store for `ByteArraysToBeRevived`.

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop/src/Microsoft.JSInterop.csproj:
```diff
@@ -8,11 +8,13 @@
     <IsAspNetCoreApp>true</IsAspNetCoreApp>
     <Nullable>enable</Nullable>
     <Trimmable>true</Trimmable>
+    <DefineConstants>$(DefineConstants);JS_INTEROP</DefineConstants>
```
TBH, given it's `internal`, I'm not really sure why we bother going through the trouble of changing the namespace in `ArrayBuilder.cs`. I know you're following on from the existing pattern with ignitor and webview, but I don't really know why they bother doing it either.

This is just to say that if you want to do this, fine, and if you don't want to bother with it, I'd support that too!

TanayParikh replied:
Thanks, think I'll keep it in that case just to stick with the general pattern elsewhere as you mentioned.

SteveSandersonMS commented:
@TanayParikh This looks really excellent and very close to being complete! Thanks so much for making such a clean job of it and being flexible about the implementation.

javiercn reviewed (commented):
Its headed in the right direction!

There are a couple of tweaks that I think are worth doing design wise, other than that I think this just needs code cleanups and tests.

I believe we need to have **TWO** methods within JSRuntime for dealing with "Serializing" and "Deserializing" byte arrays. I don't think the `DotNetDispatcher` needs to know about byte arrays.

The reasons are:
* `SupplyByteArray` on DotNetDispatcher completely delegates to the underlying JSRuntime instance and its forced to know the details about it.
* Some of the checks that we want to do there, (like the max size) are exposed to the dispatcher independent of the implementation (wasm, server, desktop) when they can be a complete implementation detail if they are handled by RemoteJsRuntime directly.

To that effect I think we need something like:
* PrepareToSerializeByteArray (Known as SupplyByteArray on IJSRuntime).
* PrepareToDeserializeByteArray (Known as SupplyByteArray on DotNetDispatcher).

The names are not great and we should find better names, the point is, JSRuntime should be the one handling these details.

javiercn commented on src/Components/Server/src/ComponentHub.cs:
```diff
@@ -209,6 +209,17 @@ public async ValueTask EndInvokeJSFromDotNet(long asyncHandle, bool succeeded, s
+                return;
+            }
+
+            await circuitHost.SupplyByteArray(id, data);
```
@pranavkm it's fine if we don't await things here. The methods in `CircuitHost` should not throw exceptions and messages will be "serialized" as they enter the sync context. That said, I don't think awaiting here has any drawback other than slightly reduced performance.

SteveSandersonMS commented:
> I believe we need to have TWO methods within JSRuntime for dealing with "Serializing" and "Deserializing" byte arrays.

Yep, sounds good! It's only a slight tweak from where this PR already is, since it already has `SupplyByteArray` on `JSRuntime`. It seems like it's just a matter of moving the `SupplyByteArray` from `DotNetDispatcher` onto `JSRuntime` and finding good names for the pair.

Suggestion: `SendByteArray` and `ReceiveByteArray`.

TanayParikh committed:
Merge branch 'main' into taparik/byteArrayDirectInterop

TanayParikh committed:
PR Feedback

TanayParikh commented:
Thanks Steve & Javier, really appreciate the feedback. I've addressed most of the feedback (https://github.com/dotnet/aspnetcore/pull/33015/commits/18ade7c0e155cfedb791814d2b4571485ce0a2c0) here. 

Still trying to get JS->.NET WASM working, proving to be a bit tricky.

javiercn commented on src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts:
```diff
@@ -394,24 +411,28 @@ export module DotNet {
+        if (jsObject) {
+          return jsObject.getWrappedObject();
+        } else {
+          throw new Error(`JS object instance with ID ${id} does not exist (has it been disposed?).`);
```
We probably don't want to be asking questions in our error messages.

javiercn replied:
We normally put quotes around identifiers like `'${id}'`

SteveSandersonMS replied:
Just for clarity - this isn't @TanayParikh's message - it comes from existing code. But we can certainly take the opportunity to clean it up 😄

TanayParikh replied:
Replacing with `JS object instance with Id '${id}' does not exist. It may have been disposed.`

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:
```diff
@@ -53,6 +56,12 @@ protected JSRuntime()
+        /// Stores the maximum SignalR message size accepted.
+        /// Used to restrict the total byte array transfers of a single call.
+        /// </summary>
+        protected long SignalRMaxMessageSizeBytes { get; set; } = 31 * 1024;
```
This needs to go in the RemoteJSRuntime directly

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:
```diff
@@ -174,6 +183,46 @@ protected virtual void BeginInvokeJS(long taskId, string identifier, string? arg
+        /// </summary>
+        /// <param name="id">Identifier for the byte array being transfered.</param>
+        /// <param name="data">Byte array to be transfered from JS.</param>
+        public void ReceiveByteArray(int id, byte[] data)
```
I think we need to have this be `virtual` so that RemoteJSRuntime can be the one doing the check about the MessageSize. We don't want to impose that limitation in other places where this doesn't matter, like WebAssembly.

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:
```diff
@@ -174,6 +183,46 @@ protected virtual void BeginInvokeJS(long taskId, string identifier, string? arg
+            {
+                throw new ArgumentOutOfRangeException($"Element id '{id}' cannot be added to the byte arrays to be revived with length '{ByteArraysToBeRevived.Count}'.");
+            }
+            else if (SignalRMaxMessageSizeBytes - data.Length < ByteArraysToBeRevivedTotalBytes)
```
I find more natural this check to be expressed as `ByteArraysToBeRevivedTotalBytes + data.Length > SignalRMessageSizeInBytes`. SignalRMessageSizeInBytes is a static limit and the other two things are the ones that change.

SteveSandersonMS replied:
The concern in previous review comments was about possible numeric overflow, hence the subtraction rather than addition.

TBH that concern seems rather artificial to me if it's a `long` (you'd have to send 8 exabytes, and as we know, nobody will even need more than 640KB).

javiercn replied:
If we are concerned about overflow we can use `checked()`. I didn't realize there was a previous comment about potential overflows and we changed it as a result.

I'm fine if we leave it as it is, I don't think it's a super powerful reason to have it done that way at the cost of readability for what you just described, however I'm fine if we don't change it.

SteveSandersonMS replied:
It wasn't me who was concerned about overflows here before, and I'm still not concerned about it 😄

pranavkm replied:
The runtime tends to use this pattern (e.g. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs#L173), so it's not really that unusual. Could we let this be?

javiercn replied:
As I said, I'm fine with it, I just haven't seen it in our code base and took me 5 seconds to get my head around it.

SteveSandersonMS commented on src/Components/Server/src/Circuits/RemoteJSRuntime.cs:
```diff
@@ -22,9 +22,14 @@ internal class RemoteJSRuntime : JSRuntime
-            _options = options.Value;
+            _options = circuitOptions.Value;
+            SignalRMaxMessageSizeBytes = hubOptions.Value.MaximumReceiveMessageSize is null ?
+                SignalRMaxMessageSizeBytes : hubOptions.Value.MaximumReceiveMessageSize.Value;
```
Apparently a `null` value is meant to mean "unlimited" ([source](https://github.com/dotnet/aspnetcore/issues/13466#issuecomment-525154822)), whereas this logic would set it to 32KB. That problem would be fixed if we changed the default in JSInterop to `long.MaxValue` as in my other comment.

javiercn replied:
@BrennanConroy What's the scenario for someone removing the limit on MaximumReceiveMessageSize. Under which circumstances do we consider that to be ok and are willing to tell customers to do so.

SteveSandersonMS replied:
Someone might have an app that doesn't even accept requests from the outside world, e.g., some kind of airgapped kiosk that just uses Blazor Server as a UI (and I know for sure one of our customers does exactly that). If SignalR already allows unlimited incoming data, I don't see any reason not to honour that API.

javiercn replied:
@SteveSandersonMS That's fine, I'm not arguing against it, we can always make any limit useless by giving it a high enough number, so its not problematic to have a way to remove it. 

I just want to get a better understanding of under which circumstances we make this recommendations because It's an improvement we can make to the Blazor Server docs.

BrennanConroy replied:
> Someone might have an app that doesn't even accept requests from the outside world, e.g., some kind of airgapped kiosk

Yeah, that's basically when we would "recommend" unlimited. When you completely trust the end user, or just don't care 😆

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:
```diff
@@ -174,6 +183,46 @@ protected virtual void BeginInvokeJS(long taskId, string identifier, string? arg
             in DotNetInvocationResult invocationResult);
 
+        /// <summary>
+        /// Invoked by JS
```
Is this comment right? I thought this was invoked by .NET code in the JSON serializer. If so, suggestion:

```suggestion
        /// Invoked by the platform to transfer a byte array from .NET to JS.
```

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:
```diff
@@ -53,6 +56,12 @@ protected JSRuntime()
+        /// Stores the maximum SignalR message size accepted.
+        /// Used to restrict the total byte array transfers of a single call.
+        /// </summary>
+        protected long SignalRMaxMessageSizeBytes { get; set; } = 31 * 1024;
```
@SteveSandersonMS See my comment below. I think this needs to be internal in RemoteJSRuntime and that we need to make `ReceiveBytes` virtual and make that check in an override within RemoteJSRuntime. That way Webassembly or other platforms don't need to know anything about this

SteveSandersonMS replied:
I think that's also difficult from a responsibility-separating point of view, as it means `RemoteJSRuntime` is now responsible for keeping track of when the data is consumed as well as when it's received.

To me it seems simpler and totally satisfactory to have a concept of a "max bytes recieved" that's platform-agnostic (and can default to `long.MaxValue`).

javiercn replied:
> I think that's also difficult from a responsibility-separating point of view, as it means `RemoteJSRuntime` is now responsible for keeping track of when the data is consumed as well as when it's received.

Not really?

I was imagining something like this on RemoteJSRuntime
```csharp
public override void ReceiveBytes(int id, byte [] data)
{
   if(id == 0) 
   {
       ReceivedBytesCount = 0;
   }

   if(ReceivedBytes + data.Length > MaxSignalrMessageSizeInBytes)
   {
       throw new InvalidOperationException("TooMuchData");
   }
   base.ReceiveBytes(int id, byte [] data);
}
```

The reason I think this is better is because it avoids additional public API surface. We've normally pushed these concepts/concerns down to the Blazor Server layer, however I'm fine if you are ok with a general purpose property and a default limit that effectively disables it. There's really not a way for anyone to set this other than producing their own implementation

SteveSandersonMS replied:
> I was imagining something like this on RemoteJSRuntime

Very good point. I hadn't thought of resetting purely based on the `id` but yes as far as I can tell that should be safe and legal (the caller can't give us more than one `id=0` value without making us discard the underlying storage anyway).

So yes, your approach looks best to me.

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:
```diff
@@ -178,15 +184,45 @@ protected virtual void BeginInvokeJS(long taskId, string identifier, string? arg
+        /// </summary>
+        /// <param name="id">Identifier for the byte array being transfered.</param>
+        /// <param name="data">Byte array to be transfered from JS.</param>
+        public void ReceiveByteArray(int id, byte[] data)
```
Ah, now I remember why we direct the calls through `DotNetDispatcher`. It's to do with access modifiers.

Really sorry we sent you in a misleading direction with an earlier comment. You were correct to have `ReceiveByteArray` initially as a `public static` method on `DotNetDispatcher`. That should then call into this method here on `JSRuntime` (which should be `internal`) to actually perform the work. This is to avoid confusing customers by having a public method on `JSRuntime` (which is a type they can legitimately work with) that they are not meant to call.

Hopefully this is only a small tweak now! Apologies for not realising this earlier. I knew there was some reason for the existing pattern but lost track of what it was.

javiercn replied:
Good catch

TanayParikh replied:
> That should then call into this method here on JSRuntime (which should be internal) 

I believe it'll have to be `protected internal` to ensure the (now) virtual method can be overridden in the `RemoteJSRuntime`. Please do let me know if you have any concerns with that 😄

SteveSandersonMS commented on src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:
```diff
@@ -174,6 +183,46 @@ protected virtual void BeginInvokeJS(long taskId, string identifier, string? arg
+            ByteArraysToBeRevived.Append(data);
+
+            // We also store the total number of bytes seen so far to compare against
+            // the max SignalR message size limit.
```
```suggestion
            // the MaximumIncomingBytes limit.
```

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,66 @@
+                throw new JsonException("ByteArraysToBeRevived is empty.");
+            }
+
+            var byteArrayRef = JsonSerializer.Deserialize<ByteArrayRef>(ref reader, options);
```
👏

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:
```diff
@@ -203,13 +203,18 @@ private static void EndInvokeDotNetAfterTask(Task task, JSRuntime jsRuntime, in
                 index++;
             }
 
+            jsRuntime.ResetByteArraysToBeRevived();
```
Once we are done de-serializing the arguments, should we check to ensure that we've consumed all incoming byte arrays and throw if that's not the case?

I'm worried that otherwise we are dropping data and I don't see any valid scenario where it would be ok to receive M byte arrays and consume only N of them during deserialization.

SteveSandersonMS replied:
It is legit to send JSON properties that aren't consumed on the receiving side. Those properties might contain an `ArrayBuffer`. It would be a further change of design to assume the .NET object shape completely covers the JS object shape, rather than allowing any subset.

I know it's *wasteful* to send properties you aren't consuming, but it would be odd to consider it an error based on the type of the property value.

javiercn replied:
Fair enough, that's a good point and a valid scenario. I think this just needs a comment then for our future selves.

TanayParikh committed:
Merge branch 'main' into taparik/byteArrayDirectInterop

TanayParikh committed:
PR Feedback

javiercn reviewed (commented):
Looks great!

I think this just needs tests now!

SteveSandersonMS commented:
Agreed - looks superb!

TanayParikh committed:
Byte Array Interop in WASM (#33104)

* Byte Array Interop in WASM

* Update src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs

TanayParikh committed:
Fix CI

TanayParikh committed:
ByteArrayJsonConverterTest.cs

TanayParikh committed:
Additional Tests

TanayParikh committed:
Add github issue

TanayParikh committed:
Update DefaultWebAssemblyJSRuntime.cs

TanayParikh committed:
Merge branch 'main' into taparik/byteArrayDirectInterop

TanayParikh committed:
Update InputFile.ts

TanayParikh commented on src/Components/Web.JS/src/InputFile.ts:
```diff
@@ -101,9 +101,9 @@ async function ensureArrayBufferReadyForSharedMemoryInterop(elem: InputElement,
 
-async function readFileData(elem: InputElement, fileId: number, startOffset: number, count: number): Promise<string> {
+async function readFileData(elem: InputElement, fileId: number, startOffset: number, count: number): Promise<Uint8Array> {
   const arrayBuffer = await getArrayBufferFromFileAsync(elem, fileId);
-  return btoa(String.fromCharCode.apply(null, new Uint8Array(arrayBuffer, startOffset, count) as unknown as number[]));
+  return new Uint8Array(arrayBuffer, startOffset, count);
 }
```
Required change to our current `InputFile` implementation as [we're expecting a `byte[]` return value](https://github.com/dotnet/aspnetcore/blob/d76ed601cab885211a73e8a90d16fd8b3f301736/src/Components/Web/src/Forms/InputFile/RemoteBrowserFileStream.cs#L60-L66).

TanayParikh committed:
Update release files

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/JSRuntime.cs:
```diff
@@ -202,7 +202,7 @@ protected internal virtual void ReceiveByteArray(int id, byte[] data)
             if (id != ByteArraysToBeRevived.Count)
             {
-                throw new ArgumentOutOfRangeException($"Element id '{id}' cannot be added to the byte arrays to be revived with length '{ByteArraysToBeRevived.Count}'.");
+                throw new ArgumentOutOfRangeException($"Element id '{id}' cannot be added to the byte arrays to be revived with length '{ByteArraysToBeRevived.Count}'.", innerException: null);
```
I don't think this is necessary? The other constructor already sets the inner exception to null. Am I missing something here?

TanayParikh replied:
So there's a slight difference between the constructors of `ArgumentOutOfRangeException(string)` and this `ArgumentOutOfRangeException(string, Exception)`. Notably, if we use the former, the parameter is the `name` of the parameter that causes this exception, and such the error message will be something like "Argument 'Element id '{id}' cannot be added to the byte arrays to be revived with length' is out of range ...`. By using the 2 parameter constructor, we can specify we want the error message to be `Element id '{id}' cannot be added to the byte arrays to be revived with length...`.

![image](https://user-images.githubusercontent.com/14852843/120524896-69a21600-c38c-11eb-9064-a3f8c1c5d6ed.png)

javiercn commented on src/JSInterop/Microsoft.JSInterop/test/JSRuntimeTest.cs:
```diff
@@ -315,6 +315,90 @@ public void CanSanitizeDotNetInteropExceptions()
+        {
+            // Arrange
+            var runtime = new TestJSRuntime();
+            runtime.DefaultTimeout = TimeSpan.FromSeconds(1);
```
Instead of doing this on every test, why not set it as the default value on the property. Other tests don't seem to be setting this.

Also you likely can do, which is normally preferred
```csharp
            var runtime = new TestJSRuntime()
            {                                     
                DefaultTimeout = TimeSpan.FromSeconds(1)
            };
```

TanayParikh replied:
Can probably just be removed, will take a look. Had copied from https://github.com/dotnet/aspnetcore/blob/1780543aa4ad0ba41dd3bc8253a8fb23ff611840/src/JSInterop/Microsoft.JSInterop/test/JSRuntimeTest.cs#L49.

javiercn commented on src/JSInterop/Microsoft.JSInterop/test/Infrastructure/ByteArrayJsonConverterTest.cs:
```diff
@@ -0,0 +1,157 @@
+        }
+
+        [Fact]
+        public void Read_ReadsJson_WithFormatting()
+        {
+            // Arrange
+            var byteArray = new byte[] { 1, 5, 7 };
+            JSRuntime.ByteArraysToBeRevived.Append(byteArray);
+
+            var json =
+@$"{{
+            ""__byte[]"": 0
+        }}";
+
+            // Act
+            var deserialized = JsonSerializer.Deserialize<byte[]>(json, JsonSerializerOptions)!;
+
+            // Assert
+            Assert.Equal(byteArray, deserialized);
+        }
```
It's good that you thought of this. That said, this is testing system.text.json which we shouldn't just do here. It's fine if we remove it.

javiercn commented on src/JSInterop/Microsoft.JSInterop/test/Infrastructure/ByteArrayJsonConverterTest.cs:
```diff
@@ -0,0 +1,157 @@
+        }
+
+        [Fact]
+        public void WriteJsonTwice_KeepsObjectId()
+        {
+            // Arrange
+            var byteArray = new byte[] { 1, 5, 7 };
+
+            // Act & Assert
+            for (var i = 0; i < 10; i++)
+            {
+                var json = JsonSerializer.Serialize(byteArray, JsonSerializerOptions);
+                Assert.Equal($"{{\"__byte[]\":{i + 1}}}", json);
+            }
+        }
```
Not sure what we are testing here, it says WriteJsonTwice but its doing something 10 times. Also the id seems to be changing every time.

I'm also not sure about the behavior here, we are not trying to de-dupe arrays, are we?

TanayParikh replied:
Thanks for pointing this out, the test name needed to be updated to `WriteJsonMultipleTimes_IncrementsByteArrayId`. 

That's correct, we don't de-dupe arrays.

javiercn reviewed (commented):
Looks great!

Just missing some E2E JS interop tests I think.

TanayParikh committed:
PR Feedback

TanayParikh committed:
Update WarningSuppression

TanayParikh committed:
E2E Tests

TanayParikh commented on src/Components/test/testassets/BasicTestApp/FormsTest/InputFileComponent.razor:
```diff
@@ -40,6 +40,8 @@ Max allowed files:
 <InputFile OnChange="LoadImage" id="input-image" />
 <br />
 
+<span id="exception-message">@exceptionMessage</span>
```
Changes in this file aren't really related to the byte array interop changes, they're more so just to provide better visibility for exceptions, to make debugging these tests easier in the future. These changes ensure `LoadImage` follows the same pattern as `LoadFiles`.

TanayParikh committed:
Merge branch 'main' into taparik/byteArrayDirectInterop

TanayParikh committed:
Update roundTripByteArrayAsyncFromJS tests to not use some ES features

TanayParikh committed:
Revert ByteArrayJsonConverter Impl

TanayParikh commented:
Had to revert the ByteArrayJsonConverter back to a previous version (https://github.com/dotnet/aspnetcore/pull/33015/commits/f710bd093f3f1913b7d9e469188d77e98ce98879) due to ongoing compatibility issues in WASM/CI. Not able to repro the issue locally, and tried logging/different solutions, but not too efficient given each attempt needs to be run through the entire CI pipeline to see if it works. 

https://github.com/dotnet/aspnetcore/pull/33015/commits/f710bd093f3f1913b7d9e469188d77e98ce98879 follows the same pattern as the JS/DotnetObjectReferenceJsonConverter so I don't think this should be a big deal.

javiercn commented on src/Components/WebAssembly/WebAssembly/src/Microsoft.AspNetCore.Components.WebAssembly.WarningSuppressions.xml:
```diff
@@ -49,23 +49,11 @@
-      <argument>IL2072</argument>
-      <property name="Scope">member</property>
-      <property name="Target">M:Microsoft.AspNetCore.Components.WebAssemblyComponentParameterDeserializer.DeserializeParameters(System.Collections.Generic.IList{Microsoft.AspNetCore.Components.ComponentParameter},System.Collections.Generic.IList{System.Object})</property>
-    </attribute>
```
@pranavkm are you happy with this change? I have no context.

TanayParikh replied:
Done based on https://github.com/dotnet/aspnetcore/tree/main/src/Components/WebAssembly/testassets/WasmLinkerTest that @pranavkm provided. Needed to fix build/CI errors.

javiercn commented on src/Components/test/testassets/BasicTestApp/InteropTest/ByteArrayInterop.cs:
```diff
@@ -0,0 +1,49 @@
+        }
+
+        [JSInvokable]
+        public static async Task<byte[]> RoundTripByteArrayAsync(byte[] byteArray)
+        {
+            await Task.Yield();
+            return byteArray;
+        }
```
No need for `async` and `Task.Yield` here. `Task.FromResult(byte)` does the trick

javiercn commented on src/Components/test/testassets/BasicTestApp/wwwroot/js/jsinteroptests.js:
```diff
@@ -219,6 +235,42 @@ function returnArrayAsync() {
 }
 
+function roundTripByteArray(byteArray) {
+  if (byteArray.constructor !== Uint8Array) {
```
I believe you can use `instanceof` here instead of checking the constructor that way.

TanayParikh replied:
I had tried going that route, but it didn't appear to be sufficient as it just defaulted to `object` instead of `Uint8Array`
![image](https://user-images.githubusercontent.com/14852843/120687706-1cd74180-c457-11eb-8d66-48a7f3debb8e.png)

javiercn commented on src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs:
```diff
@@ -0,0 +1,85 @@
+
+            int? byteArrayRef = null;
+
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
+            {
+                if (reader.TokenType == JsonTokenType.PropertyName)
+                {
+                    if (byteArrayRef is null && reader.ValueTextEquals(ByteArrayRefKey.EncodedUtf8Bytes))
+                    {
+                        reader.Read();
+                        byteArrayRef = reader.GetInt32();
+                    }
+                    else
+                    {
+                        throw new JsonException($"Unexpected JSON property {reader.GetString()}.");
+                    }
+                }
+                else
+                {
+                    throw new JsonException($"Unexpected JSON Token {reader.TokenType}.");
+                }
+            }
```
Why were we unable to use the `Serialize` trick here?

If we are parsing this in this way, can we be more strict and check for the `exact` sequence here? I believe we should be able to check something like
1. Start object
2. __byte[] property
3. End object

And for anything else we can/should fail. 

Also, if we are parsing the input like this, we need to write tests for how it behaves against malformed inputs like `{ __byte[]: "something" }`, etc.

TanayParikh replied:
> Why were we unable to use the Serialize trick here?

The `Serialize` trick causes the WASM E2E tests to fail in the CI. Not able to repro the issue locally, and tried logging/several different solutions without any luck. This isn't really efficient given each attempt needs to be run through the entire CI pipeline to see if it works.

> If we are parsing this in this way, can we be more strict and check for the exact sequence here?
> Also, if we are parsing the input like this, we need to write tests for how it behaves against malformed inputs like { __byte[]: "something" }, etc.

Sure thing, updated. 02b2b75ec1

javiercn reviewed (approved):
This looks fantastic!

I have one small comment that needs clarification. Other than that, this is good to go!

TanayParikh committed:
PR Feedback

TanayParikh committed:
Merge branch 'main' into taparik/byteArrayDirectInterop

javiercn reviewed (approved):
Good to go!

TanayParikh commented:
Breaking change announcement proposal:

# Byte Array Interop in .NET 6.0 Preview 6
### Labels:
- Breaking Change
- Enhancement
- Release

### Summary
Blazor Server & WASM now supports optimized byte array interop which avoids encoding/decoding byte arrays into Base64, facilitating a more efficient interop process.

### Breaking Change
#### Receiving Byte Arrays in JS
```js
function ReceivesByteArray(data)
{
	// Previously data was a Base 64 encoded string representing the byte array
	// 6.0 Preview 6 and beyond, it'll be a Uint8Array (no longer requires processing the Base 64 encoding)
}
```

#### Sending Byte Arrays from JS
If .NET is expecting a `byte[]` JS must provide a `Uint8Array`. Previously, it was possible to provide a Base 64 encoding array using `btoa`.

For example, if you have something like this:
```csharp
var bytes = await _jsRuntime.InvokeAsync<byte[]>("someJSMethodReturningAByteArray");
```

then you _must_ provide a `Uint8Array` from JS (must _not_ be Base 64 encoded).

**Before:**
```js
function someJSMethodReturningAByteArray() {
    const data = new Uint8Array([ 1, 2, 3 ]);
    const base64EncodedData = btoa(String.fromCharCode.apply(null, data as unknown as number[]));
    return base64EncodedData;
}
```

**After:**
```js
function someJSMethodReturningAByteArray() {
    const data = new Uint8Array([ 1, 2, 3 ]);
    return data;
}
```


PR: https://github.com/dotnet/aspnetcore/pull/33015
Issue: https://github.com/dotnet/aspnetcore/issues/21877

SteveSandersonMS commented:
Announcement looks great. Suggested tweaks:

> Blazor Server & WASM now supports optimized byte array interop

Replace WASM with WebAssembly.

> Receiving Byte Arrays in JS

For context, can you also show the corresponding .NET code that invokes this?

> Previously, it was possible to provide a Base 64 encoding array

Typo: "Base 64 encoding array" should be "Base64 encoded array"

TanayParikh commented:
Thanks @SteveSandersonMS, it's official https://github.com/dotnet/announcements/issues/187!

guardrex commented:
@TanayParikh / @SteveSandersonMS ... The announcement isn't labeled (and placing "in .NET 6.0 Preview 6" in the title isn't normal and doesn't filter well), so it's a bit hidden at the moment.

TanayParikh commented:
> @TanayParikh / @SteveSandersonMS ... The announcement isn't labeled (and placing "in .NET 6.0 Preview 6" in the title isn't normal and doesn't filter well), so it's a bit hidden at the moment. [dotnet/announcements#187](https://github.com/dotnet/announcements/issues/187)

Thanks for pointing that out @guardrex, I had tried adding labels but I don't have the adequate permissions. Do you happen to have the required permissions or know how I can get them? I'd like to add in the following labels.

- Breaking Change
- Enhancement
- Release
- .NET 6.0

Also, are you familiar with how the change goes from dotnet/announcements to https://docs.microsoft.com/dotnet/core/compatibility/6.0#aspnet-core?

guardrex commented:
I don't ... I'm just a docs :cat2:.

Open a new issue here on the PU repo. They prefer not to work from closed issues/PRs like this.

WRT the breaking change doc ... submit a docs issue (and possibly PR) for it. It looks like the docs aren't autogenerated but actually worked up manually and then modified manually. I see dev PRs going into it on the blame .....

https://github.com/dotnet/docs/blame/main/docs/core/compatibility/6.0.md

Use the **This page** button at the bottom of the topic to open the issue and go from there.
